tests/unittests/test_backoff.py

```diff
@@ -0,0 +1,209 @@
+        raise HTTPError(f"{self.status_code} Error")
+
+
+class TestOutbrainClient(unittest.TestCase):
+
+    def setUp(self):
+        self.client = OutbrainClient()
+
+    def test_rate_limit_backoff_generator_continuous(self):
+        """
+        Verify that _rate_limit_backoff yields the current retry_after value
+        repeatedly, and updates when retry_after changes.
+        """
+        gen = self.client._rate_limit_backoff()
+        first = next(gen)
+        second = next(gen)
+        self.assertEqual(first, self.client._retry_after)
+        self.assertEqual(second, self.client._retry_after)
+
+        # Changing retry_after should reflect in subsequent yields
+        self.client._retry_after = 7.25
+        third = next(gen)
+        self.assertEqual(third, 7.25)
+
+    @patch.object(SESSION, "send")
+    def test_make_request_success_with_auth_only(self, mock_send):
+        """
+        Simulate a 200 OK response when only auth is provided,
+        and assert a single call to SESSION.send.
+        """
+        resp = requests.Response()
+        resp.status_code = 200
+        resp._content = b'{"result":"ok"}'
+        resp.url = "http://auth-only/"
+        mock_send.return_value = resp
+
+        result = self.client.make_request("GET", "http://auth-only", auth=("u", "p"))
+        self.assertEqual(result.status_code, 200)
+        mock_send.assert_called_once()
+
+    @patch.object(SESSION, "send")
+    def test_make_request_success_with_headers_and_params_and_data(self, mock_send):
+        """
+        Simulate a 200 OK response when headers, params, and data are provided,
+        and verify that they appear on the prepared request.
+        """
+        resp = requests.Response()
+        resp.status_code = 200
+        resp._content = b""
+        resp.url = "http://headers-params/"
+        mock_send.return_value = resp
+
+        headers = {"X-Custom": "Val"}
+        params = {"q": "test"}
+        data = {"key": "value"}
+
+        result = self.client.make_request(
+            "post", "http://headers-params", headers=headers, params=params, data=data
+        )
+        self.assertEqual(result.status_code, 200)
+
+        sent_req = mock_send.call_args[0][0]
+        # Check headers and params are honored
+        self.assertIn("X-Custom", sent_req.headers)
+        self.assertIn("q=test", sent_req.url)
+        mock_send.assert_called_once()
+
+    @patch.object(time, "sleep", lambda s: None)
+    @patch.object(SESSION, "send")
+    def test_429_retries_n_times_then_error(self, mock_send):
+        """
+        Simulate repeated 429 responses with a valid header value,
+        verify five retry attempts, and expect Server429Error.
+        """
+        mock_send.return_value = DummyResponse(
+            429, headers={"rate-limit-msec-left": "1500"}
+        )
+
+        with self.assertRaises(Server429Error):
+            self.client.make_request("GET", "http://rate-limit/")
```

Comment:
Assert error message as well to confirm that after 5 retry, it throws an exception